README.md
# Atom-graph-database

本project包含:

1. [数据规范](docs/data_convention.md)
3. [常用名称](docs/name_list.md)
2. [数据集列表](docs/dateset_list.md)
3. [常用bash脚本](docs/scripts_list.md)
2. [数据工具（python module)](docs/datatools/introduction.md)

---

## Quick start

### 下载已有数据集

本项目使用ceph共享数据, 推荐使用sensesync进行数据的上传和下载,
参见[ceph使用说明](http://sdoc.pjlab.org.cn:10099/docs/Petrel-OSS/01-Petrel-OSS%E5%BF%AB%E9%80%9F%E5%85%A5%E9%97%A8/index.html)
和[sensesync使用说明](http://sdoc.pjlab.org.cn:10099/docs/hpc/usage/08-sensesync-backup.html).

所有本项目在ceph上的目录以agdb_xxxxx的方式命名,前缀agdb是**A**tom **G**raph **D**ata**B**ase的缩写.

在本教程中, 使用的示例数据集为`agdb_null_test`, ceph bucket名称为`agdb_null_test`. 包括基础数据(`raw`文件夹中), 两类额外文件`attr1`&`attr2`(`extra`文件夹中).

```bash
# files
agdb_null_test
|---raw  # basic file dir
|   |---001.pkl
|   |---002.pkl
|   |---...
|---extra  # extra file dir
|   |---001.attr1.pkl
|   |---001.attr2.pkl
|   |---002.attr1.pkl
|   |---002.attr2.pkl
|   |---...
|---dataset_attributes.pkl  # basic file
|---raw.tar  # tarfile of dir raw, optional
|---attr1.tar  # tarfile of extra/attr1, optional
|---attr2.tar  # tarfile of extra/attr2, optional
```

每个数据集都由数据集属性`dataset_attributes.pkl`和两个子目录`raw`和`extra`构成,
根据[数据规范](docs/data_convention.md), `dataset_attributes.pkl`中以`dict`的形式保存整个数据集的所有数据点共享的属性,
`raw`中的数据文件储存一个分子结构和基本的属性, 如能量等,
`extra`中的文件储存附加属性, 可以根据你的模型的实际需要按需下载和读取.
读取方式会在本教程的[读取本地数据集](#读取本地数据集)一节中介绍.

假设你现在已经配置好了sensesync,我们就可以使用如下命令下载数据:
``` bash
sensesync --anonymous 0 cp s3://agdb_null_test.10.140.2.204/ <your_data_path>/
```
这里你需要把`<your_data_path>`替换为一个具体的目录. 这样, 执行命令后所有数据文件就下载到了本地.

ceph有权限控制功能, 访问ceph上的目录需要相应的访问权限, 本教程中的数据集已经配置了允许[匿名访问](#上传新的数据集), 因此任何人都能通过匿名访问`--anonymous 0`获取.
在一般的情况下, 下载前你需要在[管理平台](petrel.pjlab.org.cn)上向数据集的拥有者申请访问权限.

### 读取本地数据集

假设我们已经下载了上一节中的`agdb_null_test`数据集, 我们先来看一下数据集里都有什么.
阅读本节时, 建议对照[数据规范](docs/data_convention.md)一起看, 本节的数据都遵循这一规范.

首先, 我们来看一下`raw`目录里有什么:

```bash
raw  # basic file dir
|---001.pkl
|---002.pkl
```

```python
# raw/001.pkl
>>> import pickle
>>> with open("<your_data_path>/raw/001.pkl") as f:
>>>     pickle.load(f)
{'x': array([1, 3, 5, 6]),
 'pos': array([[-1.63331094,  0.75480048,  0.9601171 ],
        [-0.47974295,  1.01928488, -1.20201389],
        [ 0.50422725,  2.32178196, -0.99859271],
        [-0.15990044, -1.33673388, -1.677831  ]]),
 'name': '001',
 'etol': -0.08012218238361365}
```

在`raw`目录中, 有两个`{name}.pkl`文件, 每个文件储存了一个分子的数据, 不同的分子用不同的`{name}`标记.

再看一下extra目录:

```bash
extra  # basic file dir
|---001.attr1.pkl
|---001.attr2.pkl
|---002.attr1.pkl
|---002.attr2.pkl
```

```python
# extra/0001.attr2.pkl
>>> with open("<your_data_path>/extra/001.attr1.pkl") as f:
>>>     pickle.load(f)
{'attr1_1': array([-1.66051672, -0.45761608,  0.5147151 , -1.85360247, -0.0466721 ]),
 'attr1_2': array([[-0.11453922, -0.09535394,  1.54399523],
        [-0.63859771,  1.11285543, -0.56919963],
        [-0.08672203,  1.29176386,  1.15865253],
        [-0.44347449, -0.98350267,  1.64920487],
        [-0.36067238, -0.17904819,  0.65143063]])}
```

可以看到,文件的格式为`{name}.{attr}.pkl`, 不同的额外属性使用`{attr}`的中间名标记.
每一个`.pkl`文件中可以储存多个属性, 读取的内容是一个`dict`:

```python
{  # the key has prefix {attr}, {} for domain deliminator, should be ignored
    '{attr}_{attr1}': ...,
    '{attr}_{attr2}': ...,
    '{attr}_{attr3}': ...,
    ...,
}
```

因此, 大多数时候我们其实不需要下载全部的extra数据, 只要下载`dataset_attributes.pkl`、全部的`raw`和需要的`extra`,
这可以通过sensesync的[正则表达式匹配](http://sdoc.pjlab.org.cn:10099/docs/Petrel-OSS/02-Petrel-OSS%E8%BF%9B%E9%98%B6%E8%AF%B4%E6%98%8E/03-sensesync%E8%BF%9B%E9%98%B6.html#id12)功能实现.

**注:接下来的内容是使用`torch_geometric`读取数据,如果你不适用这一框架,那么你需要根据数据文件的结构自己实现读取数据的代码,你可以跳到[上传新的数据集](#上传新的数据集)继续阅读**

在我们目前的工作中,常使用`torch_geometric(pyg)`来管理图网络的数据, 我们可以尝试使用`pyg`读取这个数据集.对于标准格式的数据, 得益于数据格式的标准化, 我们提供了公共的`Dataset`类来便捷地读取数据,
代码见[datatools/datasets](datatools/pyg_datasets.py)

**注: 目前datatools还不能以python module的形式安装(正在完善), 代码的使用方式是直接复制到你的文件中**

代码中实现了三个不同的数据集类:
- `Atom_graphs_dataset`: 拥有最大的灵活度
- `InMemory_Atom_graphs_dataset`: 拥有最高的读取效率
- `Direct_Atom_graphs_dataset`: 拥有最快的初始化速度

他们都继承自`torch_geometric`的[Dataset](https://pytorch-geometric.readthedocs.io/en/latest/notes/create_dataset.html), 适用于不同的场景.
训练代码中, 我们推荐优先使用`InMemory_Atom_graphs_dataset`, 更详细的说明请参见相应的代码和注释.

使用这些代码的方式非常简单, 通常只需要两个参数`root`和`extra_attrs`:

```python
>>> root = "<your_data_path>/agdb_null_test/"
>>> dataset = InMemory_Atom_graphs_dataset(root, extra_attrs=['attr1', 'attr2'])
Processing...
Done!
```

root是你的数据所在的目录,extra_attrs是需要读取的额外属性, 即extra文件的中间名, 无论是否读取多个属性, 都要放在一个`list`中提交.

第一次运行上面的代码时, 需要比较长的时间完成数据的读取和预处理, 并将处理后的数据集保存在`root/processed/InMemoryDataset.attr1_attr2.pkl`的单个文件中, 之后再进行读取就可以几乎立即完成了. 当然, 在调试代码时, 使用`Direct_Atom_graphs_dataset`会更加方便, 它能瞬间完成初始化.

```python
>>> root = "<your_data_path>/agdb_null_test/"
>>> dataset = Direct_Atom_graphs_dataset(root, extra_attrs=['attr1', 'attr2'])
```

接下来, 你就可以直接从中读取数据, 或者将它传递给一个`Dataloader`:

```python
>>> data = dataset[0]
>>> data
Data(x=[4], pos=[4, 3], name='001', etol=-0.08012218238361365, attr1_1=[5], attr1_2=[5, 3], attr2_1=[5], attr2_2=[5, 3])
```

从`dataset_attributes.pkl`中读取的数据集公共属性属性也可以从这个`dataset`中获取:

```Python
>>> dataset.dataset_attrs
{'xc': 'm062x',
 'basis': 'def2tzvp',
 'unit': {'energy': 'eV', 'length': 'Bohr'},
 'doc': 'This is a test dataset for building agdb modules'}
```

需要特别注意的是,只有存在`raw`中的数据会被数据集自动从`numpy.ndarray`转化为`torch.Tensor`,额外数据需要自己转换,例如使用下节中介绍的`transform`.

更多细节请参考[代码](datatools/pyg_datasets.py)和[pyg的文档](https://pytorch-geometric.readthedocs.io/en/latest/modules/data.html#torch_geometric.data.Dataset)

### 使用Transform

原始的数据通常不会直接作为模型的输入, 而是需要一些预处理,
比如按照一定的方式从原子的位置`pos`生成图的边`edge_index`, 在边上生成需要的属性, 如原子间的相对距离等.
这些操作我们可以通过定义需要的[transform](https://pytorch-geometric.readthedocs.io/en/latest/notes/introduction.html#data-transforms)来完成.
`transform`既可以直接作用于单个数据, 又可以作为参数传给`Dataset`使每一个数据都自动进行transform.

在`torch_geometric`中, 已经预定义了很多实用的[transform](https://pytorch-geometric.readthedocs.io/en/latest/modules/transforms.html),
我们下面尝试为`qm7_gnnxc`添加边和相对距离:

```python
>>> from torch_geometric import transforms
>>> tf1 = transforms.RadiusGraph(r=2.5)
>>> tf2 = transforms.Distance()
>>> tf = transforms.Compose([tf1,tf2])
>>> tf(data)
Data(x=[4], pos=[4, 3], name='001', etol=-0.08012218238361365, attr1_1=[5], attr1_2=[5, 3], attr2_1=[5], attr2_2=[5, 3], edge_index=[2, 6], edge_attr=[6, 1])
```

可以看到, `Data`中多了两个属性, `edge_index`和`edge_attr`. 验证transform产生了预期的效果后, 我们可以重新定义数据集:

```python
>>> dataset_with_tf = Direct_Atom_graphs_dataset(root,extra_attrs = ['attr1','attr2'],transform=tf)
>>> dataset_with_tf[0]
Data(x=[4], pos=[4, 3], name='001', etol=-0.08012218238361365, attr1_1=[5], attr1_2=[5, 3], attr2_1=[5], attr2_2=[5, 3], edge_index=[2, 6], edge_attr=[6, 1])
```
可以看到, 现在读出来的数据使已经经过transform的.

在训练过程中, 每个数据都会被反复读取, 因此在读取之后再进行transform会产生很大的额外计算成本, 更常用的方式是在初始化数据集的时候完成需要的transform, 此时我们使用`pre_transfrom`参数来定义数据集,以完成数据的预处理. 特别地, `Direct_Atom_graphs_dataset`不支持预处理, 此时我们必须使用`InMemory_Atom_graphs_dataset`或`Atom_graphs_dataset`来读取数据:
```python
>>> dataset_with_tf = Atom_graphs_dataset(root,extra_attrs = ['attr1','attr2'],pre_transform=tf)
Processing...
Done!
>>> dataset_with_tf[0]
Data(x=[4], pos=[4, 3], name='001', etol=-0.08012218238361365, attr1_1=[5], attr1_2=[5, 3], attr2_1=[5], attr2_2=[5, 3], edge_index=[2, 6], edge_attr=[6, 1])
```

`pre_transform`和`transfrom`参数可以同时存在, 如非必要情况, 比如正在debug, 或者这个操作不能够提前完成, 我们都应该优先使用`pre_transform`.

除了使用预定义的`transfrom`外, 我们也可以自己定义任意的操作. 一个常用操作是转化数据类型.
在[数据规范](docs/data_convention.md)中, 所有的数据都是用双精度(float64/int64)储存, 提高精度的代价是更大的储存空间和更慢的训练速度.

```python
>>> data.pos.dtype
torch.float64

>>> class Change_dtype(transforms.BaseTransform):
        def __init__(self,attr_names:Union[list[str],str],dtypes:Union[list[torch.dtype],torch.dtype]) -> None:
            self.attr_names = attr_names
            self.dtypes = dtypes

        def __call__(self, data: Data) -> Data:
            if isinstance(self.attr_names,str):
                self.attr_names = [self.attr_names]
            if not isinstance(self.dtypes,list):
                self.dtypes = [self.dtypes for i in range(len(self.attr_names))]
            for name,dtype in zip(self.attr_names,self.dtypes):
                attr = getattr(data,name).to(dtype=dtype)
                setattr(data,name,attr)
            return data
>>> another_tf = Change_dtype(['x','pos'],[torch.int8,torch.float32])
>>> data = another_tf(data)
>>> data.x.dtype
torch.int8
>>> data.pos.dtype
torch.float32
```

以上代码定义了一个将指定属性转化为torch.float32类型的变换. 自定义变换时, 我们需要继承`transform.BaseTransform`类, 并定义`__init__`和`__call__`两个方法. 前者是初始化该变换的实例时调用的, 后者是对数据进行变换时调用的.


**进阶说明**

pyg的dataloader和pytorch的基础dataloader的区别是定义了一个默认的`collate_fn`,使得它能够正确地生成多个`Data`数据的Batch.但是这个默认的`collate_fn`并不总能正确地处理所有类型的数据,例如当某个属性对于不同的数据是`shape`互不相同的`torch.Tensor`时,这个函数就会报错.`agdb_null_test`中的两个额外属性就是这样的.不过,如果我们不对这些属性作类型转换,保持它们为`numpy.ndarray`,默认的`collate_fn`就能正常运行.

因此,我们不能以transfrom的形式转化这些属性,因为transform发生在collate之前,相反,我们在获取数据后手动进行transform.

### 上传新的数据集


CHANGELOG.md
# Changelog

## [0.4.0] - 2023-1-17

### Added

- datatools/ceph_tools
    - `CMDClients`
    - `CEPHDataset`
- pip install support

### Changed

- rename `datatools` as `agdb`
- add cephtools introduction in `Quick start`
- bug fix in `pyg_datasets`

## [0.3.2] - 2023-01-11

### changed

- fix `Quick start`
- rename `datatools/datasets` as `datatools/pyg_datasets`
- update `datatools/pyg_datasets`
- update `data_convention`


## [0.3.1] - 2023-01-11

### Added

- prepare for packaging into a module
- add `null_test` data

### changed

- change `pyg.data.Data` and `torch.Tensor` to `dict` and `numpy.array`
- update qm7_gnnxc dataset
- kwarg `extra_attributes` to `extra_attrs` for simplicity

## [0.3] - 2023-01-10

### Added

- datatools/datasets:
    - `Atom_graphs_dataset`
    - `InMemory_Atom_graphs_dataset`
    - `Direct_Atom_graphs_dataset`
- Quick Start
    - README中的教程
    - 示例数据集`qm7_gnnxc`和`qm7_scf_M062X_6311`
    - name_list.md 和 Dataset_list.md 的示例

### Changed

- 更新了数据约定
    - 去除了部分未确定的内容
    - 额外数据的使用规则和储存方式
    - dataset的命名规则和目录结构
    - ceph的使用方式
- 删除了datatools/from_pyscf.py


## [0.2] - 2022-12-08

### Added

- 本项目框架
    - TODO 
    - Changelog
    - /docs
    - /datatools
    - /scripts

### Changed

- 更新了数据约定
    - Data中，Cell变为Optional
    - 增加数据集约定


## [0.1] - 2022-12-07

### Added

- 数据约定：Data基本数据结构

agdb/ceph_tools.py
import configparser
import datetime
import importlib
import locale
import logging
import os
import re
import shlex
import shutil
import subprocess
import tarfile
from functools import partial
from typing import Any, Dict, Generator, List, Literal, Union

from tqdm import tqdm

# if use petrel_client
USE_PETRELCLIENT:bool=None
# config petreloss
PETRELOSSCONFDOC:str = r"http://sdoc.pjlab.org.cn:10099/docs/Petrel-OSS/01-Petrel-OSS%E5%BF%AB%E9%80%9F%E5%85%A5%E9%97%A8/02-SDK%E9%85%8D%E7%BD%AE%E4%B8%8E%E7%AE%80%E5%8D%95%E6%93%8D%E4%BD%9C.html#id3"
# config aws s3
AWSS3CONFDOC:str = r"http://sdoc.pjlab.org.cn:10099/docs/Petrel-OSS/01-Petrel-OSS%E5%BF%AB%E9%80%9F%E5%85%A5%E9%97%A8/01-awscli%E9%85%8D%E7%BD%AE%E4%B8%8E%E7%AE%80%E5%8D%95%E6%93%8D%E4%BD%9C.html#id1"
# # system encoding
# ENC:str = locale.getpreferredencoding()  # get local encoding

# try to import petrel_client
try:
    import petrel_client
    from botocore.response import StreamingBody
    from petrel_client.client import Client
    from petrel_client.mixed_client import MixedClient
    USE_PETRELCLIENT = True
except ImportError:
    logging.warning("Module petrel_client not found, using CMDClient as alternative. To install petrel_client, see http://sdoc.pjlab.org.cn:10099/docs/Petrel-OSS/ for instructions")
    USE_PETRELCLIENT = False


"""
aws s3 behaviors:

ls:
- bucket
    - exists: content
    - exists / empty: nothing
    - nonexists: error, return 255
- content
    - no end `/`:
        - has dir: print dir info
        - has file: print file info
    - end with `/`:
        - has dir: print dir content info
        - no dir: error, return code 1
"""

class CMDClient:
    """provide same methods as petrel_client.client.Client
    based on aws s3 cmdlines ans regex parsing
    please refer to petrel_client.mixed_client.MixedClient for details"""

    AWS_PRE:str = "aws s3 --endpoint-url=http://10.140.2.204:80"

    PAT = {  # regex patterns
        "list": r"""
            ^([\d\-]*)\s?([\d:]*)  # datetime, optional
            \s+
            (\d+|PRE)  # size
            \s
            ([.\-\w\d]+/*)$  # fname
        """,
    }

    def __init__(self, conf_path:str=None):
        # use petreloss config or aws s3 config
        if conf_path is not None:
            logging.warning("conf_path should be None when using CMDClient, auto use ~/.s3cfg as config file")
        self.conf_path = os.path.join(os.path.expanduser("~"), ".s3cfg")
        self._check_aws_s3()
        logging.info("Using commandline-based awscli as ceph client, please make sure that awscli module is at system path")

    # def __del__(self):
    #     pass

    def _execmd(self, cmd:str, **kwargs):
        """execute a cmd and return output with decoded str"""
        if type(cmd) is str:
            cmd = shlex.split(cmd)
        elif type(cmd) is not list:
            raise TypeError(f"{cmd=} should be str or list, not {type(cmd)}")
        process = subprocess.Popen(cmd, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
        while process.poll() is None:  # wait for procedure ends
            continue
        output = process.stdout.read().decode(locale.getpreferredencoding())
        return process.returncode, output

    def _check_aws_s3(self):
        """determine endpoint-url in the conf_path
        by `aws s3 ls --endpoint-url=xxx` cmd"""
        if not os.path.exists(self.conf_path):
            raise ValueError(f"conf_path {self.conf_path} not found, please refer to {AWSS3CONFDOC} for details")

        # get endpoint config
        config = configparser.ConfigParser()
        config.read(self.conf_path, encoding=locale.getpreferredencoding())
        try:
            host_base = config.get("default", "host_base")
        except configparser.NoOptionError:
            raise ValueError(f"default/host_base not found .s3cfg, which is necessary")
        assert host_base.startswith("http://")
        self.AWS_PRE = f"aws s3 --endpoint-url={host_base}"

        # try exe a aws cmd
        cmd = f"{self.AWS_PRE} ls"
        c, s = self._execmd(cmd)
        assert c == 0, f"command `{cmd}` failed, please (re)install awscli and refer to {AWSS3CONFDOC} for details"
        return True

    def _check_uri(self, uri:str):
        """check if uri is of correct formats"""
        if not uri.startswith("s3://"):
            return False
        else:
            return True

    def _ls(self, uri:str):
        cmd = f"{self.AWS_PRE} ls {uri}"
        c, s = self._execmd(cmd)
        assert c == 0, f"command `{cmd}` failed"
        regex = re.compile(self.PAT["list"], re.VERBOSE)
        matches = [regex.search(ss) for ss in s.splitlines()]
        return [m.groups() for m in matches]  # date, time, size, fname

    def create_bucket(self, bucket:str, **kwargs):
        """create a bucket
        format of bucket: 's3://mybucket' with no ending slash"""
        assert bucket.startswith("s3://")
        assert '/' not in bucket[len("s3://"): ]
        cmd = f"{self.AWS_PRE} mb {bucket}"
        c, s = self._execmd(cmd)
        assert c == 0, f"command `{cmd}` failed"
        return c, s

    def list(self, uri:str) -> Generator:
        """list dir of this uri"""
        if not uri.endswith('/'):
            uri += '/'
        groups = self._ls(uri)
        fnames = [g[3] for g in groups]
        return fnames

    def contains(self, uri:str, **kwargs) -> bool:
        """if dir or file exists"""
        uri = uri.strip('/')
        cmd = f"{self.AWS_PRE} ls {uri}"
        c, s = self._execmd(cmd)
        if c != 0:
            return False
        else:
            return True

    def isdir(self, uri:str, **kwargs):
        """if dir exists, buckets are also dirs
        if is dir, return True
        else (e.g. non exist or file), return False"""
        assert self._check_uri(uri)
        uri = uri.rstrip("/")  # remove the right /
        if '/' not in uri[len("s3://"): ]:  # bucket
            cmd = f"{self.AWS_PRE} ls {uri}"
            c, s = self._execmd(cmd)
            if c != 0:
                return False
            else:
                return True
        else:  # dir or file
            name = uri.rsplit('/')[-1]
            uri2 = uri.rsplit('/', 1)[0]  # the parent uri
            uri2 += '/'
            groups = self._ls(uri2)
            for g in groups:
                if g[3].rstrip('/') == name:  # found
                    if g[2] == "PRE":  # is dir
                        return True
                    else:  # not dir
                        return False
            return False

    def delete(self, uri:str, **kwargs):
        """delete files"""
        raise NotImplementedError

    def get(self, uri:str, save_dir:str="./", **kwargs):
        """download files to save_dir"""
        cmd = f"{self.AWS_PRE} cp {uri} {save_dir}"
        c, s = self._execmd(cmd)
        assert c == 0, f"command `{cmd}` failed"
        return c, s


class CEPHDataset:
    """download proper files to dataset root dir from ceph"""

    # ceph_datasets.info members:
    author:str
    mail:str
    ip:str
    bucket:str
    count:int
    datafiles:Dict
    doc:str

    def __init__(self, bucket:str, conf_path:str=None):
        """load dataset info from python files of THIS name
        the bucket should exists in ceph

        Args:
            root (str): root dir of dataset, samilar meaning to pyg.data.Dataset(root)
            bucket (str): bucket name, should be exactly the same as the name of the bucket in ceph
            conf_path (str, optional): aws s3 config path. Defaults to None.
        """

        try:
            _info = importlib.import_module(f"ceph_datasets.{bucket}").info
        except:
            raise ValueError(f"Dataset {bucket} not found in ceph_dataset infomation files")
        assert _info["bucket"] == bucket
        # set info to internal attributes
        [setattr(self, k, v) for k, v in _info.items()]
        # instantialize petrel_client
        self.client = Client(conf_path=conf_path) if USE_PETRELCLIENT else CMDClient(conf_path=conf_path)
        # check if bucket exists
        assert self.client.isdir(f"s3://{bucket}"), f"bucket {bucket} not found in ceph"

    def __repr__(self):
        return f"CEPHDataset(name={self.bucket}, doc={self.doc})"

    def __doc__(self):
        return self.doc

    def get_fnames(self, attr:str) -> List[str]:
        """load single files list
        attr: None for check raw"""
        if attr == "raw":
            flist = tuple(i for i in self.client.list(f"s3://{self.bucket}/raw/") if i.endswith(".pkl"))
        elif type(attr) is str:
            extra_suffix = f".{attr}."  # the attr suffix, .[attr].
            flist = tuple(i for i in self.client.list(f"s3://{self.bucket}/extra/") if (extra_suffix in i))
        else:
            raise ValueError(f"attr should be 'raw' or str, got {attr=} with type {type(attr)}")
        if len(flist) == 0:
            raise ValueError(f"no ceph files found for {attr=}")
        return flist

    def check_bucket(self):
        """check bucket contents

        Returns:
            storage = {
                "raw": [has_single_file, has_tar_file],
                "attr1": [has_single_file, has_tar_file],
                "attr2": [has_single_file, has_tar_file],
                ...
            }
        """

        storage = {k: [None, None] for k in self.datafiles.keys()}
        for k in storage.keys():
            storage[k][0] = (self.count == len(self.get_fnames(k)))
            storage[k][1] = self.client.contains(f"s3://{self.bucket}/{k}.tar")

        # self.storage = storage
        return storage

    def _download_single(self, root:str, save_fpath:str, bucket_fpath:str, chunk_size:int=100):
        """stream download single file from bucket_fpath to save_fpath

        Args:
            root (str): dataset root to save files
            save_fpath (str): from dataset root, e.g. data/agdb_null_test/raw/001.pkl -> save_fpath='raw/001.pkl'
            bucket_fpath (str): from bucket root, e.g. bucket/raw/001.pkl -> bucket_fpath='raw/001.pkl'
            chunk_size (int, optional): chunk size in MB. Defaults to 100.

        """
        if type(self.client) is Client:
            stream:StreamingBody = self.client.get(f"s3://{self.bucket}/{bucket_fpath}", enable_stream=True)
            with open(os.path.join(root, save_fpath), 'wb') as f:
                for chunk in stream.iter_chunks(chunk_size*1024**2):
                    f.write(chunk)
        elif type(self.client) is CMDClient:
            self.client.get(f"s3://{self.bucket}/{bucket_fpath}", os.path.join(root, save_fpath))

    def download(
        self, root:str,
        extra_attrs:Union[List[str], None, Literal["all"]],
        use_tar:bool=True,
        del_tar:bool=True,
        chunk_size:int=100):
        """download data from ceph to root dir

        Args:
            root (str): root dir to the dataset
            extra_attrs (Union[List[str], None, Literal["all"]]): extra attrs to download.
                None for no extra attrs, "all" for all avaliable extra attrs.
            use_tar (bool): download tarfile instead of single files if avaliable. Defaults to True.
            del_tar (bool): delete tarfile after download. Defaults to True.
            chunk_size (int): stream download size, MB. Only effective to petrel_client. Defaults to 100.
        Returns:
            None
        """
        download_single = lambda x, y: self._download_single(root=root, save_fpath=x, bucket_fpath=y, chunk_size=chunk_size)
        global tqdm
        tqdm = partial(tqdm, ncols=100)

        check_root_dirs(root)
        extra_info = self.datafiles.copy()
        extra_info.pop("raw")
        extra_attr1s = find_extra_attr1s(extra_info, extra_attrs)  # 1 for first-level attr categories
        logging.info(f"extra_attr files: {extra_attr1s}")

        """download: 'raw' and first-level extra_attr, tar or single files, and dataset_attributes.pkl"""
        tar_fpaths = []
        # download dataset_attributes.pkl
        download_single("dataset_attributes.pkl", "dataset_attributes.pkl")

        # download raw
        if use_tar and self.client.contains(f"s3://{self.bucket}/raw.tar"):
            logging.info(f"downloading raw.tar")
            download_single("raw.tar", "raw.tar")
            tar_fpaths.append("raw.tar")
            extract_tar(os.path.join(root, "raw.tar"), root)
        else:
            fnames = self.get_fnames("raw")
            tbar = tqdm(fnames, desc="downloading raw files")
            for fname in tbar:
                download_single(os.path.join('raw', fname), os.path.join('raw', fname))

        # download extra
        for extra_attr1 in extra_attr1s:
            if use_tar and self.client.contains(f"s3://{self.bucket}/{extra_attr1}.tar"):
                logging.info(f"downloading {extra_attr1}.tar")
                download_single(os.path.join(f"{extra_attr1}.tar"), f"{extra_attr1}.tar")
                tar_fpaths.append(f"{extra_attr1}.tar")
                extract_tar(os.path.join(root, f"{extra_attr1}.tar"), root)
            else:
                fnames = self.get_fnames(extra_attr1)
                tbar = tqdm(fnames, desc=f"downloading {extra_attr1} files")
                for fname in tbar:
                    download_single(os.path.join('extra', fname), os.path.join('extra', fname))

        # remove tar files
        if del_tar:
            [os.remove(os.path.join(root, f)) for f in tar_fpaths]

        logging.info(f"Done!")

    def upload(self, root:str, use_tar:bool=True, test:bool=False):
        """upload local data"""
        raise NotImplementedError

    def fly(self, root:str):
        """load data on the fly
        consider mount ceph to the root dir"""
        raise NotImplementedError



def extract_tar(fpath_tar:str, extract_dir:str):
    """extract tar file to root dir"""
    try:
        tarobj = tarfile.open(fpath_tar)
    except tarfile.TarError:
        raise OSError(
            "%s is not a compressed or uncompressed tar file" % fpath_tar)
    try:
        tarobj.extractall(extract_dir)
    finally:
        tarobj.close()


def check_root_dirs(root:str):
    """check if root, raw, extra dir exists, or create them"""
    if not os.path.exists(root):
        os.mkdir(root)
        logging.info(f"Created dir {root}")
    for d in ["raw", "extra"]:
        dir_path = os.path.join(root, d)
        if not os.path.exists(dir_path):
            os.mkdir(dir_path)
            logging.info(f"Created dir {dir_path}")


def find_extra_attr1s(extra_info:Dict[str, List[str]], extra_attrs:Union[List[str], None, Literal["all"]]) -> List[str]:
    """find first level extra files' names by comparing extra_info and extra_attrs

    Args:
        extra_info (Dict[str, List[str]]): information about dataset extra attrs.
        extra_attrs (Union[List[str], None, Literal["all"]]): required extra attributes, first-level or second-level attrs.
            None for no extra attrs, "all" for all avaliable extra attrs.
    Returns:
        List[str]: extra file names
    """
    if extra_attrs is None:
        return []
    elif (type(extra_attrs) == str) and (extra_attrs == "all"):
        return list(extra_info.keys())

    # common situations
    assert type(extra_attrs) is list
    extra_fnames = []
    for extra_attr in extra_attrs:
        found_attr = False  # record if found attr
        # iterate over the first level and second level attrs
        for attr1, attr2s in extra_info.items():
            if (extra_attr == attr1) or (extra_attr in attr2s):
                extra_fnames.append(attr1)
                found_attr = True
                continue
        if not found_attr:
            raise ValueError(f"extra attr {extra_attr} not found in dataset, with avaliable extra attrs {extra_info}")

    extra_fnames = list(set(extra_fnames))  # eliminate multiplicity
    return extra_fnames


def _test2():
    logging.basicConfig(level=logging.INFO, format='%(asctime)s %(message)s',)

    root = "/nvme/louzekun/Downloads/agdb_qm7_gnnxc"
    if os.path.exists(root):
        shutil.rmtree(root)
        logging.info(f"delete root dir {root}")

    ceph_dataset = CEPHDataset(bucket="agdb_qm7_gnnxc")
    logging.info(ceph_dataset.__repr__())
    logging.info(ceph_dataset.__doc__())
    ceph_dataset.check_bucket()
    ceph_dataset.download(root=root, use_tar=True, extra_attrs=["scf_rho"], chunk_size=100)


def _test1():
    logging.basicConfig(level=logging.INFO, format='%(asctime)s %(message)s',)

    root = "/nvme/louzekun/atom-graph-database/datatools/ceph_datasets/agdb_null_test_downloads"
    if os.path.exists(root):
        shutil.rmtree(root)
        logging.info(f"delete root dir {root}")

    ceph_dataset = CEPHDataset(bucket="agdb_null_test")
    logging.info(ceph_dataset.__repr__())
    logging.info(ceph_dataset.__doc__())
    ceph_dataset.check_bucket()
    ceph_dataset.download(root=root, use_tar=False, extra_attrs="all", chunk_size=100)
    ceph_dataset.download(root=root, use_tar=True, extra_attrs="all", chunk_size=100)

    # os.rmdir(root)

def _test2():
    global USE_PETRELCLIENT
    USE_PETRELCLIENT = False
    logging.basicConfig(level=logging.INFO, format='%(asctime)s %(message)s',)

    root = "/nvme/louzekun/atom-graph-database/datatools/ceph_datasets/agdb_null_test_downloads"
    if os.path.exists(root):
        shutil.rmtree(root)
        logging.info(f"delete root dir {root}")

    ceph_dataset = CEPHDataset(bucket="agdb_null_test")
    logging.info(ceph_dataset.__repr__())
    logging.info(ceph_dataset.__doc__())
    ceph_dataset.check_bucket()
    ceph_dataset.download(root=root, use_tar=False, extra_attrs="all", chunk_size=100)
    ceph_dataset.download(root=root, use_tar=True, extra_attrs="all", chunk_size=100)


if __name__ == "__main__":
    _test1()
    _test2()


